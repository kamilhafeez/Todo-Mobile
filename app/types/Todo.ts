export interface Todo {
  _id: string;
  sessionId: string;
  title: string;
  completed: boolean;
}
