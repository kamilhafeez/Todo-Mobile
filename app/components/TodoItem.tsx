import React, { useCallback } from "react";
import { TouchableOpacity } from "react-native";
import styled from "styled-components/native";
import { Todo } from "@/app/types/Todo";

interface TodoItemProps {
  todo: Todo;
  onEdit: (todo: Todo) => void;
  onDelete: (id: string) => void;
  onToggleComplete: (id: string) => void;
}

const TodoItem: React.FC<TodoItemProps> = ({
  todo,
  onEdit,
  onDelete,
  onToggleComplete,
}) => {
  const onToggleHandler = useCallback(
    () => onToggleComplete(todo._id),
    [todo, onToggleComplete]
  );
  const onEditHandler = useCallback(() => onEdit(todo), [todo, onEdit]);
  const onDeleteHandler = useCallback(
    () => onDelete(todo._id),
    [todo, onDelete]
  );
  return (
    <Container>
      <CheckboxContainer>
        <TouchableOpacity testID="checkbox" onPress={onToggleHandler}>
          <Checkbox completed={todo.completed} />
        </TouchableOpacity>
      </CheckboxContainer>

      <TodoText completed={todo.completed}>{todo.title}</TodoText>
      <ButtonContainer>
        <StyledButton title="Edit" onPress={onEditHandler} />
        <StyledButton title="Delete" onPress={onDeleteHandler} />
      </ButtonContainer>
    </Container>
  );
};

export default TodoItem;

const Container = styled.View`
  flex-direction: row;
  align-items: center;
  margin-vertical: 5px;
`;

const CheckboxContainer = styled.View`
  margin-right: 10px;
`;

const Checkbox = styled.View<{ completed: boolean }>`
  width: 20px;
  height: 20px;
  border: 2px solid ${(props) => (props.completed ? "#4caf50" : "#ccc")};
  background-color: ${(props) => (props.completed ? "#4caf50" : "transparent")};
  border-radius: 3px;
`;

const TodoText = styled.Text<{ completed: boolean }>`
  flex: 1;
  font-size: 16px;
  text-decoration: ${(props) => (props.completed ? "line-through" : "none")};
`;

const ButtonContainer = styled.View`
  flex-direction: row;
  gap: 10px;
`;

const StyledButton = styled.Button``;
