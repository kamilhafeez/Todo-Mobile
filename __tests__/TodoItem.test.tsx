import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import TodoItem from "@/app/components/TodoItem";
import { Todo } from "@/app/types/Todo";

describe("TodoItem Component", () => {
  const mockTodo: Todo = {
    _id: "1",
    sessionId: "xyz",
    title: "Test Todo",
    completed: false,
  };

  const mockOnEdit = jest.fn();
  const mockOnDelete = jest.fn();
  const mockOnToggleComplete = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders the todo item correctly", () => {
    const { getByText } = render(
      <TodoItem
        todo={mockTodo}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
      />
    );

    expect(getByText("Test Todo")).toBeTruthy();
    expect(getByText("Edit")).toBeTruthy();
    expect(getByText("Delete")).toBeTruthy();
  });

  it("calls onToggleComplete when the checkbox is clicked", () => {
    const { getByTestId } = render(
      <TodoItem
        todo={mockTodo}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
      />
    );

    const checkbox = getByTestId("checkbox");
    fireEvent.press(checkbox);

    expect(mockOnToggleComplete).toHaveBeenCalledWith("1");
  });

  it("calls onEdit when the Edit button is pressed", () => {
    const { getByText } = render(
      <TodoItem
        todo={mockTodo}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
      />
    );

    const editButton = getByText("Edit");
    fireEvent.press(editButton);

    expect(mockOnEdit).toHaveBeenCalledWith(mockTodo);
  });

  it("calls onDelete when the Delete button is pressed", () => {
    const { getByText } = render(
      <TodoItem
        todo={mockTodo}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
      />
    );

    const deleteButton = getByText("Delete");
    fireEvent.press(deleteButton);

    expect(mockOnDelete).toHaveBeenCalledWith("1");
  });

  it("renders the todo text with strike-through when completed", () => {
    const completedTodo = { ...mockTodo, completed: true };
    const { getByText } = render(
      <TodoItem
        todo={completedTodo}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
      />
    );

    const todoText = getByText("Test Todo");
    expect(todoText.props.style).toEqual(
      expect.objectContaining({ textDecorationLine: "line-through" })
    );
  });

  it("renders the checkbox with correct background color when completed", () => {
    const completedTodo = { ...mockTodo, completed: true };
    const { toJSON } = render(
      <TodoItem
        todo={completedTodo}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        onToggleComplete={mockOnToggleComplete}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
