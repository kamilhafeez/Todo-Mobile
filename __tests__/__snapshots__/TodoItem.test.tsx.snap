// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem Component renders the checkbox with correct background color when completed 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "marginVertical": 5,
    }
  }
>
  <View
    style={
      {
        "marginRight": 10,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="checkbox"
    >
      <View
        completed={true}
        style={
          {
            "backgroundColor": "#4caf50",
            "borderBottomLeftRadius": 3,
            "borderBottomRightRadius": 3,
            "borderColor": "#4caf50",
            "borderStyle": "solid",
            "borderTopLeftRadius": 3,
            "borderTopRightRadius": 3,
            "borderWidth": 2,
            "height": 20,
            "width": 20,
          }
        }
      />
    </View>
  </View>
  <Text
    completed={true}
    style={
      {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "fontSize": 16,
        "textDecorationColor": "black",
        "textDecorationLine": "line-through",
        "textDecorationStyle": "solid",
      }
    }
  >
    Test Todo
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
        "gap": 10,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Edit
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Delete
        </Text>
      </View>
    </View>
  </View>
</View>
`;
